libchisel/src/binaryenopt.rs

```diff
@@ -0,0 +1,107 @@
+/*
+ * Binaryen optimiser.
+ */
+
+use std::collections::HashMap;
+
+use super::{ModuleError, ModulePreset, ModuleTranslator};
+use parity_wasm::elements::*;
+
+// FIXME: change level names
+pub enum BinaryenOptimiser {
+    O0, // Baseline aka no changes
+    O1,
+    O2,
+    O3,
+    O4,
+    Os,
+    Oz,
+}
+
+impl ModulePreset for BinaryenOptimiser {
+    fn with_preset(preset: &str) -> Result<Self, ()> {
+        match preset {
+            "O0" => Ok(BinaryenOptimiser::O0),
+            "O1" => Ok(BinaryenOptimiser::O1),
+            "O2" => Ok(BinaryenOptimiser::O2),
+            "O3" => Ok(BinaryenOptimiser::O3),
+            "Os" => Ok(BinaryenOptimiser::Os),
+            "Oz" => Ok(BinaryenOptimiser::Oz),
+            _ => Err(()),
+        }
+    }
+}
+
+impl ModuleTranslator for BinaryenOptimiser {
+    fn translate_inplace(&self, module: &mut Module) -> Result<bool, ModuleError> {
+        Err(ModuleError::NotSupported)
+    }
+
+    fn translate(&self, module: &Module) -> Result<Module, ModuleError> {
+        // FIXME: could just move this into `BinaryenOptimiser`
+        let config = match self {
+            O0 => binaryen::CodegenConfig {
+                optimization_level: 0,
+                shrink_level: 0,
+            },
+            O1 => binaryen::CodegenConfig {
+                optimization_level: 1,
+                shrink_level: 0,
+            },
+            O2 => binaryen::CodegenConfig {
+                optimization_level: 2,
+                shrink_level: 0,
+            },
+            O3 => binaryen::CodegenConfig {
+                optimization_level: 3,
+                shrink_level: 0,
+            },
+            Os => binaryen::CodegenConfig {
+                optimization_level: 2,
+                shrink_level: 1,
+            },
+            Oz => binaryen::CodegenConfig {
+                optimization_level: 2,
+                shrink_level: 2,
+            },
+        };
+
+        // FIXME: there must be a better way to accomplish this.
+        let serialised = parity_wasm::elements::serialize::<Module>(module.clone())
+            .expect("invalid input module");
+        let output = binaryen_optimiser(&serialised, &config);
+        Ok(
+            parity_wasm::elements::deserialize_buffer::<Module>(&output[..])
+                .expect("invalid output module"),
+        )
+    }
+}
+
+fn binaryen_optimiser(input: &[u8], config: &binaryen::CodegenConfig) -> Vec<u8> {
+    // NOTE: this can abort (panic) if the input is invalid
+    let module = binaryen::Module::read(&input);
+    // NOTE: this is a global setting...
+    binaryen::set_global_codegen_config(&config);
```

Comment:
Oh, I just realized that this API might be not sound...

`set_global_codegen_config` uses a mutex to set global optimization parameters. However, these are read by functions like `BinaryenModuleOptimize` without any locking. My C-Memory-model-fu is not good, but I guess there is no happens-before edge between `BinaryenModuleOptimize` and `BinaryenSetOptimizeLevel` and mutex locking only on one side is no help.

I think this issue might be exposed if you use `wasm-opt` with different codegen configs with multiple threads.

Reply:
Yeah found a couple of issues in binaryen-rs 😉